src/cluster_legacy.c

```diff
@@ -5986,6 +6226,245 @@ char *clusterNodeGetShardId(clusterNode *node) {
+        return false;
+    }
+
+    *slot = s;
+    *node = n;
+    *timeout_ms = t;
+    return true;
+}
+
+void clusterSetSlotCommand(client *c) {
+    int slot;
+    int timeout_ms;
+    clusterNode *n;
+
+    if (!clusterParseSetSlotCommand(c, &slot, &n, &timeout_ms)) return;
+
+    /* Enhance cluster topology change resilience against primary failures by
+     * replicating SETSLOT before execution.
+     *
+     * Cluster topology changes such slot ownership and migrating states must
+     * be replicated to replicas before applying them to the primary. This
+     * guarantees that after a command is successfully executed, the new state
+     * won't be lost due to a primary node failure. The following example
+     * illustrates how a cluster state can be lost during slot ownership
+     * finalization:
+     *
+     * When finalizing the slot, the target primary node B might send a cluster
+     * PONG to the source primary node A before the SETSLOT command is replicated
+     * to replica node B'. If primary node B crashes at this point, B' will be in
+     * the importing state and the slot will have no owner.
+     *
+     * To mitigate this issue, the following order needs to be enforced for slot
+     * migration finalization such that the replicas finalize the slot ownership
+     * before the primary:
+ .   *
+     * 1. Client C issues SETSLOT n NODE B against node B.
+     * 2. Primary B replicates `SETSLOT n NODE B` to all of its replicas (e.g., B', B'').
+     * 3. Upon replication completion, primary B executes `SETSLOT n NODE B` and
+     *    returns success to client C.
+     * 4. The following steps can happen in parallel:
+     *   a. Client C issues `SETSLOT n NODE B` against parimary A.
+     *   b. Primary B gossips its new slot ownership to the cluster (including A, A', etc.).
+     *
+     * This ensures that all replicas have the latest topology information, enabling
+     * a reliable slot ownership transfer even if the primary node went down during
+     * the process. */
+    if (nodeIsMaster(myself) && myself->numslaves != 0 && (c->flags & CLIENT_PREREPL_DONE) == 0) {
+        forceCommandPropagation(c, PROPAGATE_REPL);
+        /* We are a primary and this is the first time we see this `SETSLOT`
+         * command. Force-replicate the command to all of our replicas
+         * first and only on success will we handle the command.
+         * Note that
+         * 1. All replicas are expected to ack the replication within the given timeout
+         * 2. The repl offset target is set to the master's current repl offset + 1.
+         *    There is no concern of partial replication because replicas always
+         *    ack the repl offset at the command boundary. */
+        if (timeout_ms == 0) {
+            timeout_ms = CLUSTER_OPERATION_TIMEOUT;
+        }
+        blockForPreReplication(c, mstime()+timeout_ms, server.master_repl_offset+1, myself->numslaves);
+        replicationRequestAckFromSlaves();
```

Comment:
So it's not possible to override this by calling `CLUSTER SETSLOT` with `TIMEOUT 0` then? I'm fine with this behaviour. I just think it's worth noting.

Reply:
skipping the wait essentially means disabling this feature. I am not sure why we would like to allow it but I am open to a discussion. is there a use case you can think of? 

> I just think it's worth noting.
Will do

Reply:
> is there a use case you can think of?

no

Reply:
If a node is disconnected might be a valid use case? If a replica is unavailable for some reason, you might still want this to through. Maybe you're trying to move the data from an unhealthy to a healthy shard by migrating the data.

Reply:
Yeah, anyway if you do `TIMEOUT 1` it waits 1ms but if you do `TIMEOUT 0` it waits a few seconds. Perhaps not what you'd expect.

Reply:
It's better that TIMEOUT 0 doesn't wait at all. It's more logical. Follow-up issue?

Reply:
Follow-up #422 opened to track this. Feel free to resolve and merge without it.